<a href="https://colab.research.google.com/github/Ritvik-Bhatia/EDA-HotelBookingAnalysis/blob/main/Hotel_Booking_Analysis_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## <b> Have you ever wondered when the best time of year to book a hotel room is? Or the optimal length of stay in order to get the best daily rate? What if you wanted to predict whether or not a hotel was likely to receive a disproportionately high number of special requests? This hotel booking dataset can help you explore those questions!

## <b>This data set contains booking information for a city hotel and a resort hotel, and includes information such as when the booking was made, length of stay, the number of adults, children, and/or babies, and the number of available parking spaces, among other things. All personally identifying information has been removed from the data. </b>

## <b> Explore and analyze the data to discover important factors that govern the bookings. </b>

# 1. Importing necessary libraries to be used in the project


In [1]:
# libraries to load the dataset and to perfrom analysis on the dataset
import pandas as pd
import numpy as np

# libraries for visualization
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns



# 2. Mounting the drive to access the dataset

## Mounting the drive to access the dataset

In [65]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## Importing Data from Drive

In [66]:
file_path = '/content/drive/MyDrive/AlmaBetter/capstone_projects/EDA_on_Hotel_Price_Booking/Hotel Bookings.csv'
hoteldata = pd.read_csv(file_path)


# 3. Basic Analisys and Filteration of the dataset


## checking the first five observations of the dataset by using .head() method.

In [67]:
hoteldata.head()

hotel  is_canceled  ...  reservation_status  reservation_status_date
0  Resort Hotel            0  ...           Check-Out               2015-07-01
1  Resort Hotel            0  ...           Check-Out               2015-07-01
2  Resort Hotel            0  ...           Check-Out               2015-07-02
3  Resort Hotel            0  ...           Check-Out               2015-07-02
4  Resort Hotel            0  ...           Check-Out               2015-07-03

[5 rows x 32 columns]

## checking the last five observations of the dataset by using .tail() method.

In [68]:
hoteldata.tail()

hotel  is_canceled  ...  reservation_status  reservation_status_date
119385  City Hotel            0  ...           Check-Out               2017-09-06
119386  City Hotel            0  ...           Check-Out               2017-09-07
119387  City Hotel            0  ...           Check-Out               2017-09-07
119388  City Hotel            0  ...           Check-Out               2017-09-07
119389  City Hotel            0  ...           Check-Out               2017-09-07

[5 rows x 32 columns]

## checking the shape of the dataset by using .shape method.

In [69]:
hoteldata.shape

(119390, 32)

By using .shape method, we could tell that this Dataset contains a huge number of booking entries with quite a proper amount of details. We have more than 1 lakh booking enteries that we can tell by looking at the number of rows with detailed description that are spread across 32 different features.

## checking the list of indexes to get a basic understanding of Dataset using .columns method. 

In [70]:
hoteldata.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='object')

## checking some basic information of data in Dataset and details about Data types in different columns of the dataset using .info() method.


In [71]:
hoteldata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

## checking the missing-values in the dataset and displaying the percentage of data missing in the columns.


In [72]:
hoteldata.isnull().sum()[hoteldata.isnull().sum()!=0]/len(hoteldata)*100

children     0.003350
country      0.408744
agent       13.686238
company     94.306893
dtype: float64

Observations after looking at the datatypes and null values:


There are one lakh nineteen thousand three hundred and ninety rows in the dataset which means there are 119390 entries to 32 different indexes.

The columns of the dataset are of different datatypes mostly "object", "integer" or "float" type. Also we can see and observe that a few columns in the dataset have missing values. We need to deal with them by either not using those indexes or removing them from the database.

The float type columns are - "children", "company", "agent" and "adr" but according to my observation this data must have been of Int datatype as number of children cannot be a float so looking at the data we found that there are some null values which makes this data - Float as float datatype supports the storage of NaN values. 

Out of these four columns :

  1. children - contains information about number of children in the dataset.
  
  2. company - contains information about company's ID responsible for making the booking or the payment maybe. in this index almost 94% of the values are missing so this data connot provide us any help in making any descision.

  3. agent - might contain the information about the agent code who has done the booking, looking at the data we can guess that around 87% of bookings are made by some agents and rest 13.7%(null values) are either self made bookings or agent name not recorded.

  4. adr - Index Name dosen't suggest anything so looking into the values and googling suggests - adr stands for "Average Daily Rate" which is defined by dividing the sum of all lodging transactions by the total number of staying nights.


While dealing with the null-values of these columns, we will also change their datatypes to integer for simplicity.

## Renaming the column adr to average_daily_rate for simplicity


In [73]:
hoteldata.rename(columns={"adr":"average_daily_rate"}, inplace = True)

In [74]:
# checking if the change is successful.
hoteldata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

## Let's drop the column company which has around 95% of the observations as missing and printing the dataset.

In [75]:
hoteldata.drop("company", axis=1, inplace=True)
hoteldata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 31 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

## Checking the unique values in the column "agent" and assuming that all null values in agent column(13.7%) are bookings that are self made without the interfarance of any agent. Therefore changing these NaN values to "0" which for our referance will be the bookings made by self also this will make data precise and without any null values.


In [76]:
hoteldata["agent"].sort_values().unique()

array([  1.,   2.,   3.,   4.,   5.,   6.,   7.,   8.,   9.,  10.,  11.,
        12.,  13.,  14.,  15.,  16.,  17.,  19.,  20.,  21.,  22.,  23.,
        24.,  25.,  26.,  27.,  28.,  29.,  30.,  31.,  32.,  33.,  34.,
        35.,  36.,  37.,  38.,  39.,  40.,  41.,  42.,  44.,  45.,  47.,
        50.,  52.,  53.,  54.,  55.,  56.,  57.,  58.,  59.,  60.,  61.,
        63.,  64.,  66.,  67.,  68.,  69.,  70.,  71.,  72.,  73.,  74.,
        75.,  77.,  78.,  79.,  81.,  82.,  83.,  85.,  86.,  87.,  88.,
        89.,  90.,  91.,  92.,  93.,  94.,  95.,  96.,  98.,  99., 103.,
       104., 105., 106., 107., 110., 111., 112., 114., 115., 117., 118.,
       119., 121., 122., 126., 127., 128., 129., 132., 133., 134., 135.,
       138., 139., 141., 142., 143., 144., 146., 147., 148., 149., 150.,
       151., 152., 153., 154., 155., 156., 157., 158., 159., 162., 163.,
       165., 167., 168., 170., 171., 173., 174., 175., 177., 179., 180.,
       181., 182., 183., 184., 185., 187., 191., 19

In [77]:
hoteldata["agent"].fillna(0, inplace = True)

converting the values in agent to integer from float.

In [78]:
hoteldata = hoteldata.astype({'agent' : 'int'})

In [79]:
hoteldata.isnull().sum()[hoteldata.isnull().sum()!=0]/len(hoteldata)*100

children    0.003350
country     0.408744
dtype: float64

## Checking the unique values in the column "children" and assuming that all 4 null values in children column(0.003%) are misplaced values in bookings therefore changing these NaN values to median of this children values which will make data precise and without any null values.

In [80]:
hoteldata["children"].value_counts()

0.0     110796
1.0       4861
2.0       3652
3.0         76
10.0         1
Name: children, dtype: int64

In [81]:
hoteldata["children"].fillna(hoteldata["children"].median(), inplace = True)

converting the values in children to integer from float.

In [82]:
hoteldata = hoteldata.astype({"children" : "int"})

In [83]:
hoteldata["children"].value_counts()

0     110800
1       4861
2       3652
3         76
10         1
Name: children, dtype: int64

## Finally Checking the unique values in column "country" and assuming that all null values in country column(0.4%) are values misplaced during bookings due to some error by the agent or staff members. Therefore changing these NaN values to most occuring values assuming that the values were not inserted as was common which will make data precise and without any null values.


In [84]:
hoteldata["country"].value_counts()

PRT    48590
GBR    12129
FRA    10415
ESP     8568
DEU     7287
       ...  
FJI        1
SLE        1
VGB        1
ASM        1
NAM        1
Name: country, Length: 177, dtype: int64

In [85]:
hoteldata['country'].fillna(hoteldata['country'].mode()[0], inplace = True)

this could be done using top member from value_counts() method but that would be a more complex way therefore I chose to use .mode() method instead, also [0] after .mode() points to the value at top.

In [86]:
hoteldata.isnull().sum()[hoteldata.isnull().sum()!=0]/len(hoteldata)*100

Series([], dtype: float64)

In [87]:
hoteldata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 31 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119390 non-null  int64  
 11  babies                          119390 non-null  int64  
 12  meal            

hence Dataset has no null values

## checking the statistical summary of the dataset using .describe() method.

In [88]:
hoteldata.describe()

is_canceled  ...  total_of_special_requests
count  119390.000000  ...              119390.000000
mean        0.370416  ...                   0.571363
std         0.482918  ...                   0.792798
min         0.000000  ...                   0.000000
25%         0.000000  ...                   0.000000
50%         0.000000  ...                   0.000000
75%         1.000000  ...                   1.000000
max         1.000000  ...                   5.000000

[8 rows x 19 columns]

 **Observation after looking at statistical summary:**

  1. The index average_daily_rate have minimum value to be negative(-6.38%) which is not possible as average price cannot be negative.
  2. The average cancelations are 37.04% and it deviates by 48.29% indicating lots of variations between cancelation.
  3. looking at the special request column, we found that around 75% of the customers asked for one special request.

## checking the observations corresponding to the negative average daily rate

In [89]:
hoteldata[hoteldata['average_daily_rate']<0]

hotel  is_canceled  ...  reservation_status  reservation_status_date
14969  Resort Hotel            0  ...           Check-Out               2017-03-15

[1 rows x 31 columns]


As there is only one record deviating tthe data, we can drop that perticular record.

In [90]:
hoteldata.drop(hoteldata[hoteldata['average_daily_rate']<0].index, inplace=True)

checking for the shape of the dataset now.


In [91]:
hoteldata.shape

(119389, 31)

# 4. Data Visualisation